README.md
# audioplayerui


Audio Player UI Plugin.


## Features
* Play (url / local file)
* Resume / pause

## Installation

First, add `audioplayerui` as a dependency in your `pubspec.yaml` file. 

Then add the AudioPlayerView like this 

```dart
  AudioPlayerController audioPlayerController = AudioPlayerController();

  @override
  Widget build(BuildContext context) {
    return AudioPlayerView(
              audioPlayerController: audioPlayerController,
              trackUrl: "",
              isLocal: false,
              trackTitle: "",
              trackSubtitle: "",
              imageUrl: ""
              );
  }
``` 

## Available methods for AudioPlayerController
| Method        | description           | parameters  | notes |
| ------------- |:-------------:| -----:|-----:|
| play      | Plays a audio file from an remote link | url ||
| playLocal      | Plays a local audio file |  path ||
| pause      | Pause's the currently playing track      |    ||
| resume |  Resumes the currently paused track      |     ||
| stop |  Stops the playback of the current track     |    ||

## Example

Demonstrates how to use the audioplayerui plugin.

See the [example documentation](example/README.md) for more information.

## Contributing

Feel free to contribute by opening issues and/or pull requests. Your feedback is very welcome!

## Changelog

See [CHANGELOG.md](CHANGELOG.md).

## License

MIT License

Copyright (c) [2019] [Joran Dob]
Copyright (c) [2019] [QREATE]




CHANGELOG.md
## [0.0.7] - 25/11/2019
* refactored code a bit 
* added new design option with boolean simple design

## [0.0.6] - 12/11/2019

* Added example 
* Updated completed status for button

## [0.0.5] - 07/11/2019

* Added reset after track playback completed.

## [0.0.4] - 02/10/2019

* Added autoplay property to controller.

## [0.0.3] - 27/08/2019

* Fixed color for pause button. 

## [0.0.2] - 23/08/2019

* Implemented basic ui and functions
* Resolved some issues

## [0.0.1] - TODO: Add release date.

* TODO: Describe initial release.


test/audioplayerui_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:audioplayerui/audioplayerui.dart';

void main() {
//  test('adds one to input values', () {
//    final calculator = Calculator();
//    expect(calculator.addOne(2), 3);
//    expect(calculator.addOne(-7), -6);
//    expect(calculator.addOne(0), 1);
//    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//  });
}


example/lib/main.dart
import 'package:audioplayerui/audioplayerui.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;
  AudioPlayerController audioPlayerController = AudioPlayerController();

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            AudioPlayerView(
              audioPlayerController: audioPlayerController,
              trackTitle: "test",
              trackUrl:
                  "https://d11b76aq44vj33.cloudfront.net/media/645/audio/5adeecf6e08d8.mp3",
              simpleDesign: false,
            )
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/ui/defaultUi.dart
import 'package:audioplayers/audioplayers.dart';
import 'package:flutter/material.dart';

class DefaultUi extends StatelessWidget {
  final AudioPlayer audioPlayer;
  final String imageUrl;
  final String trackTitle;
  final String trackSubtitle;
  final bool hasNext;
  final bool hasPrevious;
  final String trackPosition;
  final String trackLength;
  final double playbackPosition;
  final AudioPlayerState audioPlayerState;
  final Function(double position) seekTrack;

  const DefaultUi({
    Key key,
    this.audioPlayer,
    this.imageUrl,
    this.trackTitle,
    this.trackSubtitle,
    this.hasNext = false,
    this.hasPrevious = false,
    this.playbackPosition = 0.0,
    this.audioPlayerState = AudioPlayerState.STOPPED,
    this.trackPosition = "00:00",
    this.trackLength = "00:00",
    this.seekTrack,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final ThemeData theme = Theme.of(context);
    final TextStyle descriptionStyle = theme.textTheme.subhead;

    return Card(
        elevation: 6,
        shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.all(Radius.circular(12.0))),
        color: Colors.white,
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: <Widget>[
            // Photo and title.

            imageUrl != null
                ? Column(
                    children: <Widget>[
                      Image.network(imageUrl),
                      Divider(),
                    ],
                  )
                : Container(
                    child: null,
                  ),
            DefaultTextStyle(
              softWrap: false,
              overflow: TextOverflow.ellipsis,
              style: descriptionStyle,
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: <Widget>[
                  Padding(
                    padding:
                        const EdgeInsets.only(left: 8.0, right: 8.0, top: 8),
                    child: Row(
                      crossAxisAlignment: CrossAxisAlignment.center,
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: <Widget>[
                        trackTitle != null || trackSubtitle != null
                            ? Expanded(
                                child: Padding(
                                padding:
                                    const EdgeInsets.only(left: 18, top: 10),
                                child: Column(
                                  crossAxisAlignment: CrossAxisAlignment.start,
                                  children: <Widget>[
                                    trackTitle != null
                                        ? Text(
                                            trackTitle,
                                            style: TextStyle(
                                                fontSize: 18,
                                                fontWeight: FontWeight.w500),
                                          )
                                        : Container(
                                            child: null,
                                          ),
                                    trackSubtitle != null
                                        ? Text(trackSubtitle)
                                        : Container(
                                            child: null,
                                          ),
                                  ],
                                ),
                              ))
                            : Container(
                                child: null,
                              ),
                        Container(
                          child: Padding(
                            padding:
                                const EdgeInsets.only(top: 18.0, right: 18),
                            child: Row(
                              crossAxisAlignment: CrossAxisAlignment.center,
                              mainAxisAlignment: MainAxisAlignment.end,
                              children: <Widget>[
                                hasPrevious
                                    ? IconButton(
                                        icon: Icon(Icons.skip_previous),
                                        onPressed: () {},
                                      )
                                    : Container(
                                        child: null,
                                      ),
                                audioPlayerState == AudioPlayerState.STOPPED ||
                                        audioPlayerState ==
                                            AudioPlayerState.PAUSED ||
                                        audioPlayerState ==
                                            AudioPlayerState.COMPLETED
                                    ? FloatingActionButton(
                                        onPressed: () {
                                          audioPlayer.resume();
                                        },
                                        tooltip: 'Play',
                                        backgroundColor: theme.accentColor,
                                        child: Icon(
                                          Icons.play_arrow,
                                          color: Colors.white,
                                        ),
                                        mini: true,
                                      )
                                    : FloatingActionButton(
                                        onPressed: () {
                                          audioPlayer.pause();
                                        },
                                        tooltip: 'Pause',
                                        backgroundColor: theme.accentColor,
                                        child: Icon(
                                          Icons.pause,
                                          color: Colors.white,
                                        ),
                                        mini: true,
                                      ),
                                hasNext
                                    ? IconButton(
                                        icon: Icon(Icons.skip_next),
                                        onPressed: () {},
                                      )
                                    : Container(
                                        child: null,
                                      ),
                              ],
                            ),
                          ),
                        )
                      ],
                    ),
                  ),
                  Column(
                    children: <Widget>[
                      Padding(
                        padding: const EdgeInsets.only(
                          left: 8.0,
                          right: 8.0,
                          top: 0.0,
                        ),
                        child: Slider(
                          onChanged: seekTrack,
                          value: playbackPosition,
                          activeColor: theme.accentColor,
                          inactiveColor:
                              Color.alphaBlend(theme.accentColor, Colors.white),
                        ),
                      ),
                      Padding(
                        padding: const EdgeInsets.only(
                          left: 28.0,
                          right: 28.0,
                          bottom: 12.0,
                        ),
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.spaceBetween,
                          children: <Widget>[
                            Text(
                              trackPosition,
                              style: TextStyle(fontSize: 12),
                            ),
                            Text(
                              trackLength,
                              style: TextStyle(fontSize: 12),
                            ),
                          ],
                        ),
                      )
                    ],
                  ),
                ],
              ),
            ),
          ],
        ));
  }
}


lib/audioplayerui.dart
library audioplayerui;

import 'package:audioplayers/audioplayers.dart';
import 'package:audioplayerui/ui/defaultUi.dart';
import 'package:audioplayerui/ui/simpleUi.dart';
import 'package:flutter/material.dart';

class AudioPlayerController {
  AudioPlayer audioPlayer = AudioPlayer();

  final bool autoPlay;

  AudioPlayerController({this.autoPlay = false});

  void play(String url) {
    audioPlayer.play(url);
  }

  void playLocal(String localPath) {
    audioPlayer.play(localPath, isLocal: true);
  }

  void pause() {
    audioPlayer.pause();
  }

  void stop() {
    audioPlayer.stop();
  }

  void resume() {
    audioPlayer.resume();
  }
}

class AudioPlayerView extends StatefulWidget {
  final AudioPlayerController audioPlayerController;
  final String trackTitle;
  final String trackSubtitle;
  final String trackUrl;
  final bool isLocal;
  final String imageUrl;
  final bool simpleDesign;

  const AudioPlayerView(
      {Key key,
      @required this.audioPlayerController,
      this.trackTitle,
      this.trackSubtitle,
      this.trackUrl,
      this.isLocal = false,
      this.imageUrl,
      this.simpleDesign = false})
      : super(key: key);

  @override
  _AudioPlayerViewState createState() => _AudioPlayerViewState(
        audioPlayerController,
        trackTitle,
        trackSubtitle,
        this.trackUrl,
        this.isLocal,
        this.imageUrl,
        this.simpleDesign,
      );
}

class _AudioPlayerViewState extends State<AudioPlayerView> {
  Duration duration = Duration();
  final AudioPlayerController audioPlayerController;
  AudioPlayer audioPlayer;
  bool hasNext = false;
  bool hasPrevious = false;
  double playbackPosition = 0.0;
  AudioPlayerState audioPlayerState = AudioPlayerState.STOPPED;
  final String trackTitle;
  final String trackSubtitle;
  final String trackUrl;
  final bool isLocal;
  final String imageUrl;
  final bool simpleDesign;

  //
  String trackPosition = "00:00";
  String trackLength = "00:00";

  _AudioPlayerViewState(
    this.audioPlayerController,
    this.trackTitle,
    this.trackSubtitle,
    this.trackUrl,
    this.isLocal,
    this.imageUrl,
    this.simpleDesign,
  );

  @override
  void initState() {
    audioPlayer = audioPlayerController.audioPlayer;
    if (audioPlayerController.autoPlay) {
      _playTrack();
    } else {
      _initTrackPlayback();
    }
    _initPositionChangeListener();
    _initTrackChangeListener();

    super.initState();
  }

  _initTrackPlayback() async {
    await audioPlayer.setUrl(trackUrl, isLocal: isLocal).then((val) {
      audioPlayer.getDuration().then((duration) {
        Duration audioPlayerDuration = Duration(milliseconds: duration);
        print("audioPlayerDuration $duration");
        setState(() {
          trackLength = _printDuration(audioPlayerDuration);
        });
      });
    });
    print(trackUrl);
  }

  _playTrack() {
    audioPlayer.play(trackUrl, isLocal: isLocal);
  }

  _initPositionChangeListener() async {
    audioPlayer.onAudioPositionChanged.listen((Duration p) async {
      if (p.inMilliseconds ==
          Duration(milliseconds: await audioPlayer.getDuration())
              .inMilliseconds) {
        audioPlayer.seek(Duration(seconds: 0));
        audioPlayer.stop();
      }
      Duration audioPlayerDuration =
          Duration(milliseconds: await audioPlayer.getDuration());
      int trackDuration = ((await audioPlayer.getDuration()) / 1000).round();
      String trackLengthFormat = _printDuration(audioPlayerDuration);
      String trackPositionFormat = _printDuration(p);
      setState(() {
        playbackPosition = (p.inSeconds / trackDuration).toDouble();
        trackLength = trackLengthFormat;
        trackPosition = trackPositionFormat;
      });
    });
  }

  String _printDuration(Duration duration) {
    String twoDigits(int n) {
      if (n >= 10) return "$n";
      return "0$n";
    }

    String twoDigitMinutes = twoDigits(duration.inMinutes.remainder(60));
    String twoDigitSeconds = twoDigits(duration.inSeconds.remainder(60));
    return "$twoDigitMinutes:$twoDigitSeconds";
  }

  _initTrackChangeListener() {
    audioPlayer.onPlayerStateChanged
        .listen((AudioPlayerState audioPlayerStateUpdate) {
      setState(() {
        audioPlayerState = audioPlayerStateUpdate;
      });
    });
  }

  _seekTrack(double position) async {
    int trackDuration = ((await audioPlayer.getDuration()) / 1000).round();
    int seekPosition = (position * trackDuration).round();
    audioPlayer.seek(Duration(seconds: seekPosition));
  }

  @override
  Widget build(BuildContext context) {
    return simpleDesign
        ? SimpleUi(
            audioPlayer: audioPlayer,
            imageUrl: imageUrl,
            trackTitle: trackTitle,
            trackSubtitle: trackSubtitle,
            hasNext: hasNext,
            hasPrevious: hasPrevious,
            playbackPosition: playbackPosition,
            audioPlayerState: audioPlayerState,
            trackPosition: trackPosition,
            trackLength: trackLength,
            seekTrack: _seekTrack,
          )
        : DefaultUi(
            audioPlayer: audioPlayer,
            imageUrl: imageUrl,
            trackTitle: trackTitle,
            trackSubtitle: trackSubtitle,
            hasNext: hasNext,
            hasPrevious: hasPrevious,
            playbackPosition: playbackPosition,
            audioPlayerState: audioPlayerState,
            trackPosition: trackPosition,
            trackLength: trackLength,
            seekTrack: _seekTrack,
          );
  }

  @override
  void dispose() {
    audioPlayer.release();
    audioPlayer.dispose();
    super.dispose();
  }
}


lib/ui/simpleUi.dart
import 'package:audioplayers/audioplayers.dart';
import 'package:flutter/material.dart';

class SimpleUi extends StatelessWidget {
  final AudioPlayer audioPlayer;
  final String imageUrl;
  final String trackTitle;
  final String trackSubtitle;
  final bool hasNext;
  final bool hasPrevious;
  final String trackPosition;
  final String trackLength;
  final double playbackPosition;
  final AudioPlayerState audioPlayerState;
  final Function(double position) seekTrack;

  const SimpleUi(
      {Key key,
      this.audioPlayer,
      this.imageUrl,
      this.trackTitle,
      this.trackSubtitle,
      this.hasNext,
      this.hasPrevious,
      this.trackPosition,
      this.trackLength,
      this.playbackPosition,
      this.audioPlayerState,
      this.seekTrack})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    final ThemeData theme = Theme.of(context);

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: <Widget>[
        Card(
          elevation: 6,
          shape: RoundedRectangleBorder(
              borderRadius: BorderRadius.all(Radius.circular(12.0))),
          color: Color.fromRGBO(21, 27, 33, 1.0),
          child: Padding(
            padding: const EdgeInsets.all(12.0),
            child: Row(
              children: <Widget>[
                audioPlayerState == AudioPlayerState.STOPPED ||
                        audioPlayerState == AudioPlayerState.PAUSED ||
                        audioPlayerState == AudioPlayerState.COMPLETED
                    ? FloatingActionButton(
                        onPressed: () {
                          audioPlayer.resume();
                        },
                        tooltip: 'Play',
                        backgroundColor: theme.accentColor,
                        child: Icon(
                          Icons.play_arrow,
                          color: Colors.white,
                        ),
                        mini: true,
                      )
                    : FloatingActionButton(
                        onPressed: () {
                          audioPlayer.pause();
                        },
                        tooltip: 'Pause',
                        backgroundColor: theme.accentColor,
                        child: Icon(
                          Icons.pause,
                          color: Colors.white,
                        ),
                        mini: true,
                      ),
                Expanded(
                  child: Column(
                    children: <Widget>[
                      Padding(
                        padding: const EdgeInsets.only(
                          left: 8.0,
                          right: 8.0,
                          top: 0.0,
                        ),
                        child: SliderTheme(
                          data: SliderThemeData(
                              trackShape: RoundedRectSliderTrackShape()),
                          child: Slider(
                            onChanged: seekTrack,
                            value: playbackPosition,
                            activeColor: theme.accentColor,
                            inactiveColor: Colors.white,
                          ),
                        ),
                      ),
                      Padding(
                        padding: const EdgeInsets.only(
                          left: 28.0,
                          right: 28.0,
                          bottom: 12.0,
                        ),
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.spaceBetween,
                          children: <Widget>[
                            Text(
                              trackPosition,
                              style:
                                  TextStyle(fontSize: 12, color: Colors.white),
                            ),
                            Text(
                              trackLength,
                              style:
                                  TextStyle(fontSize: 12, color: Colors.white),
                            ),
                          ],
                        ),
                      )
                    ],
                  ),
                ),
              ],
            ),
          ),
        ),
        Padding(
          padding: const EdgeInsets.only(left: 12.0),
          child: Text(
            trackTitle,
            style: TextStyle(
                color: Color.fromRGBO(188, 189, 193, 1.0),
                fontStyle: FontStyle.italic),
          ),
        ),
      ],
    );
  }
}
